File: back/src/main/java/br/unitins/tp2/service/ProdutoService.java
package br.unitins.tp2.service;

import java.util.List;

import br.unitins.tp2.dto.ProdutoDTO;
import br.unitins.tp2.model.Produto;
import jakarta.validation.Valid;

public interface ProdutoService {

    Produto create(@Valid ProdutoDTO dto);
    Produto update(long id, @Valid ProdutoDTO dto);
    void delete(long id);
    Produto findById(long id);
    List<Produto> findAll(int page, int pageSize);
    List<Produto> findByNome(String nome, int page, int pageSize);
    long count();
    long count(String nome);
}

File: front/src/app/produtos/produto.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { inject, Injectable } from '@angular/core';
import { Observable } from 'rxjs';

import { PageResponse } from '../shared/page-response.model';
import { Produto, ProdutoPayload } from './produto.model';

@Injectable({ providedIn: 'root' })
export class ProdutoService {
  private readonly http = inject(HttpClient);
  private readonly url = 'http://localhost:8080/produtos';

  findAll(nome = '', page = 0, pageSize = 5): Observable<PageResponse<Produto>> {
    const endpoint = nome.trim() ? `${this.url}/nome/${encodeURIComponent(nome.trim())}` : this.url;
    const params = new HttpParams().set('page', page).set('pageSize', pageSize);
    return this.http.get<PageResponse<Produto>>(endpoint, { params });
  }

  findById(id: number): Observable<Produto> {
    return this.http.get<Produto>(`${this.url}/${id}`);
  }

  create(payload: ProdutoPayload): Observable<Produto> {
    return this.http.post<Produto>(this.url, payload);
  }

  update(id: number, payload: ProdutoPayload): Observable<Produto> {
    return this.http.put<Produto>(`${this.url}/${id}`, payload);
  }

  delete(id: number): Observable<void> {
    return this.http.delete<void>(`${this.url}/${id}`);
  }
}
